pkg/pgmodel/end_to_end_tests/write_test.go

```diff
@@ -0,0 +1,241 @@
+package end_to_end_tests
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/route"
+	"github.com/timescale/timescale-prometheus/pkg/api"
+	"github.com/timescale/timescale-prometheus/pkg/internal/testhelpers"
+	"github.com/timescale/timescale-prometheus/pkg/pgmodel"
+	"github.com/timescale/timescale-prometheus/pkg/prompb"
+	"github.com/timescale/timescale-prometheus/pkg/util"
+)
+
+type dataGenerator struct {
+	metricGroup int //nonoverlapping metrics
+	queueID     int
+
+	metricsPerGroup    int
+	labelSetsPerMetric int
+	maxSamples         int64
+	deltaTime          int64
+	deltaValue         float64
+	currentTime        int64
+}
+
+func (t *dataGenerator) generateTimeseries() []prompb.TimeSeries {
+	metrics := []prompb.TimeSeries{}
+	for metric := 0; metric < t.metricsPerGroup; metric++ {
+		for instance := 0; instance < t.labelSetsPerMetric; instance++ {
+			labelSet := []prompb.Label{
+				{Name: pgmodel.MetricNameLabelName, Value: fmt.Sprintf("metric_%d_%d", t.metricGroup, metric)},
+				{Name: "foo", Value: fmt.Sprintf("bar_%d", t.queueID)}, //queues have non-overlapping label sets
+				{Name: "instance", Value: fmt.Sprintf("%d", instance)},
+			}
+
+			if instance%2 == 0 {
+				labelSet = append(labelSet,
+					prompb.Label{Name: "sometimes_label", Value: "constant"})
+			}
+
+			metrics = append(metrics, prompb.TimeSeries{
+				Labels: labelSet,
+			})
+		}
+
+	}
+
+	numSamples := rand.Int63n(t.maxSamples)
+	for i := range metrics {
+		metrics[i].Samples = t.generateSamples(numSamples)
+	}
+	t.currentTime += t.maxSamples*t.deltaTime + int64(1)
+	return metrics
+}
+
+func (t *dataGenerator) generateSamples(count int64) []prompb.Sample {
+	samples := make([]prompb.Sample, 0, 3)
+	i := int64(0)
+	for i < count {
+		samples = append(samples, prompb.Sample{
+			Timestamp: t.currentTime + (t.deltaTime * int64(i)),
+			Value:     float64(t.metricGroup*10000) + (t.deltaValue * float64(i)),
+		})
+		i++
+	}
+
+	return samples
+}
+
+func getHTTPWriteRequest(protoRequest *prompb.WriteRequest) (*http.Request, error) {
+	data, err := proto.Marshal(protoRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	body := string(snappy.Encode(nil, data))
+	u, err := url.Parse(fmt.Sprintf("http://%s:%d/write", testhelpers.PromHost, testhelpers.PromPort.Int()))
+
+	if err != nil {
+		return nil, err
+	}
+	return http.NewRequest(
+		"POST",
+		u.String(),
+		strings.NewReader(body),
+	)
+}
+
+func sendWriteRequest(t testing.TB, router *route.Router, ts []prompb.TimeSeries) {
+	req, err := getHTTPWriteRequest(&prompb.WriteRequest{Timeseries: ts})
```

Comment:
Should we use our pool here as well?

Reply:
I don't think so since this is the "client" generating the request and it wouldn't use the same pool or be on the same machine